src/main/kotlin/org/rust/cargo/project/model/impl/CargoProjectImpl.kt

```diff
@@ -314,6 +313,15 @@ data class CargoProjectImpl(
         "CargoProject(manifest = $manifest)"
 }
 
+private inline fun AtomicReference<VirtualFile>.getCachedFile(ioFileGenerator: () -> File?): VirtualFile? {
+    val cached = get()
+    if (cached != null && cached.isValid) return cached
+    val ioFile = ioFileGenerator()
+    val file = if (ioFile != null) LocalFileSystem.getInstance().findFileByIoFile(ioFile) else null
+    set(file)
+    return file
+}
```

Comment:
FWIW, I am totally unsure if this hand-rolled caching is valid :-) There are some `VirtualFilePointerSomething` classes in IntelliJ which do something similar. I've investigated them, but the results were inconclusive.